### PR TITLE
Preserve the original component type in merging to the array

### PR DESCRIPTION
**Pull request by yurkom**
It fixes an exception:

```
Exception in thread "main" com.fasterxml.jackson.databind.JsonMappingException: 
Problem deserializing property 'date1' (
expected type: [array type, component type: [simple type, class java.util.Date]];
actual type: `java.lang.Object[]`),
problem: Can not set [Ljava.util.Date; field com.example.Data.dates to [Ljava.lang.Object; 
(through reference chain: com.example.Data["date2"])
```

### CODE REVIEW

**Comment by yurkom** [comment]
It would be nice if this fix will be backported to 2.14 and 2.15 also.

**Comment by cowtowncoder** [comment]
Looks good, thank you for contributing this @yurkom !

One thing before merging (and considering where to backport): unless we have a CLA from you for earlier PRs (if so, please LMK), we'd need one now:

https://github.com/FasterXML/jackson/blob/master/contributor-agreement.pdf

(this is only needed once before the first contribution to any Jackson project)

and is usually done by printing doc, filling & signing, scan/photo, email to `info` at fasterxml dot com.

Once this is received I can merge PR.

As to backporting, since this looks quite safe it could go in 2.15 for sure (for which there will probably eventually be a patch release); I could merge it in 2.14 but it's not clear if or when a patch might be released.
Still, could add there in case of eventual release.

**Comment by cowtowncoder** [comment]
Yes, re-creating PR to target 2.14 would be nice. But I can alternatively just cherry-pick this one too if that's too much hassle.
I think going to 2.14 is fine here.

I can merge this in 2 days (right now can't commit but can comment :) )

**Review by cowtowncoder — APPROVED** [review]

**Comment by yurkom** [comment]
Hi,
I have submitted the CLA and created 2 pull requests for 2.14 (#4121) and 2.15 (#4120).

**Comment by cowtowncoder** [comment]
> Hi, I have submitted the CLA and created 2 pull requests for 2.14 (#4121) and 2.15 (#4120).

Thank you; apologies for not mentioning that it is enough to create just one (oldest branch, 2.14) and we can merge forward.

I will merge close this issue and leave 2.14 pr; hope to merge some time tomorrow or day after (right now don't have commit access due traveling)

**Comment by cowtowncoder** [comment]
Replaced by #4121

**Pull request closed without merging**
